tokenclient/tokenFlows.go

```diff
@@ -0,0 +1,145 @@
+	"encoding/json"
+	"fmt"
+	"github.com/sap/cloud-security-client-go/env"
+	"github.com/sap/cloud-security-client-go/httpclient"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// Options allows configuration http(s) client
+type Options struct {
+	HTTPClient *http.Client // Default: basic http.Client with a timeout of 10 seconds and allowing 50 idle connections
+}
+
+// RequestOptions allows to configure the token request
+type RequestOptions struct {
+	// Request parameters that shall be overwritten or added to the payload
+	Params map[string]string
+}
+
+// TokenFlows setup once per application.
+type TokenFlows struct {
+	identity env.Identity
+	options  Options
+	tokenURI string
+}
+
+// RequestFailedError represents a HTTP server error
+type RequestFailedError struct {
+	// StatusCode of failed request
+	StatusCode int
+	url        url.URL
+	errTxt     string
+}
+
+// Error initializes RequestFailedError
+func (e *RequestFailedError) Error() string {
+	return fmt.Sprintf("request to '%v' failed with status code '%v' and payload: '%v'", e.url.String(), e.StatusCode, e.errTxt)
+}
+
+type tokenResponse struct {
+	Token string `json:"access_token"`
+}
+
+const (
+	tokenEndpoint              string = "/oauth2/token" //nolint:gosec
+	grantTypeParameter         string = "grant_type"
+	grantTypeClientCredentials string = "client_credentials"
+	clientIDParameter          string = "client_id"
+	clientSecretParameter      string = "client_secret"
+)
+
+// NewTokenFlows initializes token flows
+//
+// identity provides credentials and url to authenticate client with identity service
+// options specifies rest client including tls config.
+// Note: Setup of default tls config is not supported for windows os. Module crypto/x509 supports SystemCertPool with go 1.18 (https://go-review.googlesource.com/c/go/+/353589/)
+func NewTokenFlows(identity env.Identity, options Options) (*TokenFlows, error) {
+	t := TokenFlows{
+		identity: identity,
+		tokenURI: identity.GetURL() + tokenEndpoint,
+		options:  options,
+	}
+	if options.HTTPClient == nil {
+		tlsConfig, err := httpclient.DefaultTLSConfig(identity)
+		if err != nil {
+			return nil, err
+		}
+		t.options.HTTPClient = httpclient.DefaultHTTPClient(tlsConfig)
+	}
+	return &t, nil
+}
+
+// ClientCredentials implements the client credentials flow (RFC 6749, section 4.4).
+// Clients obtain an access token outside of the context of a user.
+// It is used for non interactive applications (a CLI, a batch job, or for service-2-service communication) where the token is issued to the application itself,
+// instead of an end user for accessing resources without principal propagation.
+//
+// ctx carries the request context like the deadline or other values that should be shared across API boundaries.
+// customerTenantURL like "https://custom.accounts400.ondemand.com" gives the host of the customers ias tenant
+// options allows to provide a request context and optionally additional request parameters
```

Comment:
Actually this comment looks like a leftover from when the context was part of the struct.
The rest lgtm